common/messaging/kafka/client_impl_test.go

```diff
@@ -0,0 +1,159 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package kafka
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/tally"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log/testlogger"
+	"github.com/uber/cadence/common/metrics"
+)
+
+func TestNewKafkaClient(t *testing.T) {
+	metricsClient := metrics.NewClient(tally.NoopScope, metrics.History)
+	logger := testlogger.New(t)
+	testCases := []struct {
+		name        string
+		config      *config.KafkaConfig
+		checkApp    bool
+		hasErr      bool
+		expectedErr string
+	}{
+		{
+			name: "Missing clusters",
+			config: &config.KafkaConfig{
+				Clusters: map[string]config.ClusterConfig{},
+			},
+			checkApp:    true,
+			hasErr:      true,
+			expectedErr: "Empty Kafka Cluster Config",
+		},
+		{
+			name: "Missing topics",
+			config: &config.KafkaConfig{
+				Clusters: map[string]config.ClusterConfig{
+					"testCluster": {
+						Brokers: []string{"testBrokers"},
+					},
+				},
+				Topics: map[string]config.TopicConfig{},
+			},
+			checkApp:    true,
+			hasErr:      true,
+			expectedErr: "Empty Topics Config",
+		},
+		{
+			name: "Missing Applications",
+			config: &config.KafkaConfig{
+				Clusters: map[string]config.ClusterConfig{
+					"test-cluster": {
+						Brokers: []string{"test-brokers"},
+					},
+				},
+				Topics: map[string]config.TopicConfig{
+					"test-topic": {
+						Cluster: "test-cluster",
+					},
+				},
+				Applications: map[string]config.TopicList{},
+			},
+			checkApp:    true,
+			hasErr:      true,
+			expectedErr: "Empty Applications Config",
+		},
+		{
+			name: "Missing topics config",
+			config: &config.KafkaConfig{
+				Clusters: map[string]config.ClusterConfig{
+					"test-cluster": {
+						Brokers: []string{"test-brokers"},
+					},
+				},
+				Topics: map[string]config.TopicConfig{
+					"test-topic": {
+						Cluster: "test-cluster",
+					},
+				},
+				Applications: map[string]config.TopicList{
+					"test-app": {
+						Topic:    "test-topic",
+						DLQTopic: "test-topic-dlq",
+					},
+				},
+			},
+			checkApp:    true,
+			hasErr:      true,
+			expectedErr: "Missing Topic Config for Topic test-topic-dlq",
+		},
+		{
+			name: "Normal Case",
+			config: &config.KafkaConfig{
+				Clusters: map[string]config.ClusterConfig{
+					"test-cluster": {
+						Brokers: []string{"test-brokers"},
+					},
+				},
+				Topics: map[string]config.TopicConfig{
+					"test-topic": {
+						Cluster: "test-cluster",
+					},
+					"test-topic-dlq": {
+						Cluster: "test-cluster",
+					},
+				},
+				Applications: map[string]config.TopicList{
+					"test-app": {
+						Topic:    "test-topic",
+						DLQTopic: "test-topic-dlq",
+					},
+				},
+			},
+			checkApp: true,
+			hasErr:   false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if !tc.hasErr {
```

Comment:
you can get rid of this if/else and always apply the defer func. within defer func you can check if an err is expected

Reply:
Thanks for catching this, updated!